### ISSUE TITLE
Using Boolean Placeholder for is_training Causes Problems when Loading on iOS

### ISSUE THREAD

**Issue by cancan101** [tensorflow/tensorflow#6124]
Currently I am training a model that uses Batch Normalization and Dropout. Because of this I build the model to take a placeholder, `is_training`, that I can feed_dict true or false depending on whether the graph is being used for training or validation.

I take the trained model and export using `tensorflow.contrib.session_bundle.exporter` however when I try to load this graph on iOS, I run into problems (see [SO](http://stackoverflow.com/questions/40855271/no-opkernel-was-registered-to-support-op-switch-with-these-attrs-on-ios)):
```
No OpKernel was registered to support Op 'Switch' with these attrs.
```
I have tried removing the placeholder and injecting the actual value of `False` for `is_training` (see [SO](http://stackoverflow.com/questions/40852729/permanently-inject-constant-into-tensorflow-graph-for-inference) and [GH](https://github.com/tensorflow/tensorflow/issues/5919)), however the model still does not load (see [SO](http://stackoverflow.com/questions/40852729/permanently-inject-constant-into-tensorflow-graph-for-inference) post):
```
Input 0 of node dropout6/cond/dropout/random_uniform/max was passed bool from dropout6/cond/Switch:1 incompatible with expected INVALID.
```

It seems like something is missing to clean up / recompile the graph, stripping out the training part of the graph because the [inject](https://github.com/tensorflow/tensorflow/issues/5919) does not seem like enough. This was one idea to solve the problem: https://github.com/tensorflow/tensorflow/issues/5919.

Right now, I can instead of using placeholder for `is_training` use constants and create the model twice, sharing weights, and then export just the validation subgraph and then load this into iOS, but the downside is that I need to instantiate the model twice into the graph.

EDIT: Would be nice to not have to resort to compiling iOS with switch: https://github.com/tensorflow/tensorflow/issues/2680#issuecomment-237483554 since that node shouldn't even be needed in the final graph. It would be nice to fully remove it from the saved graph.

**Comment by ronny3050** [comment]
I struggled a lot with this until I gave up and ended up freezing the model without is_training.

**Comment by petewarden** [comment]
Sorry we weren't able to give you a better solution for this one! We do generally recommend having separate graphs for training and inference, though that isn't well documented. Closing this bug since it looks like you have a workaround? Let me know if we should reopen it.

**Comment by zaneoliver6** [comment]
Hey guys! I've been having the same problem trying to make this work on android and I have used some of the solutions from above but none seem to work...here are some of the errors I get back from the various methods of trying to solve this problem...Any Help would be appreciated!

The models work when I run it on python but doesn't when I run them on android.

Im using the SSD_Mobilenet model from tensorflow object detection api.

This is from the optimize for inference function:
NodeDef expected inputs '' do not match 1 inputs specified; Op<name=Const; signature= -> output:dtype; attr=value:tensor; attr=dtype:type>; NodeDef: Preprocessor/map/while/add/y = Constdtype=DT_INT32, value=Tensor<type: int32 shape: [] values: 1>

This is from renaming Switch to Identify:
Not a valid TensorFlow Graph serialization: Node 'Preprocessor/map/while/add/y': Connecting to invalid output 1 of source node Preprocessor/map/while/Switch which has 1 outputs

This last one loads the model but crashes when processing the inputs:
java.lang.IllegalArgumentException: No OpKernel was registered to support Op 'Switch' with these attrs. Registered devices: [CPU], Registered kernels:
device='GPU'; T in [DT_STRING]
device='GPU'; T in [DT_BOOL]
device='GPU'; T in [DT_INT32]
device='GPU'; T in [DT_FLOAT]
device='CPU'; T in [DT_FLOAT]
device='CPU'; T in [DT_INT32]

                                                              	 [[Node: Postprocessor/BatchMultiClassNonMaxSuppression/PadOrClipBoxList/cond/Switch = Switch[T=DT_BOOL](Postprocessor/BatchMultiClassNonMaxSuppression/PadOrClipBoxList/Greater, Postprocessor/BatchMultiClassNonMaxSuppression/PadOrClipBoxList/Greater)]]
Any ideas guys? I've also tried to recompile tensorflow with the changes from [https://stackoverflow.com/questions/40855271/no-opkernel-was-registered-to-support-op-switch-with-these-attrs-on-ios/43627334#43627334](url) but bazel crashes while compiling :(

**Comment by lili25** [comment]
@ronny3050 could you share how to freeze the model is_training ? 
thanks.

**Comment by ronny3050** [comment]
[https://gist.github.com/ronny3050/d3dc669601a5604b347365ed652645dd](This gist) can be used for freezing models. Do not include `is_training` in the list of whitelisted nodes. 
You may have to change a couple of hardcorded node-names to fit your needs.

Hope it helps. :)

**Comment by lili25** [comment]
@ronny3050, thanks for your reply, this fixed my issue.

**Comment by ronny3050** [comment]
Just FYI, we can simply feed phase_train and keep_prob etc. along with input in Android and it fixes the issue.

``tensorflow.feed("phase_train", new boolean[]{false});
    tensorflow.feed("keep_prob", new float[]{1f});
``

**Comment by snehabhapkar10** [comment]
@ronny3050 If you know can you please tell me, how to make phase_train option false if I have .pb(frozen model).

**Comment by iamyb** [comment]
I wrote an example below to convert a tensorflow(v1.3) model with batchnorm to tflite. Hope it's helpful.
https://gist.github.com/iamyb/779a14c2da79c7d7fe5d7256e4d71ace
